BaseClasses.py

```diff
@@ -427,28 +425,22 @@ def push_item(self, location: Location, item: Item, collect: bool = True):
 
         logging.debug('Placed %s at %s', item, location)
 
-    def get_entrances(self) -> List[Entrance]:
-        if self._cached_entrances is None:
-            self._cached_entrances = [entrance for region in self.regions for entrance in region.entrances]
-        return self._cached_entrances
-
-    def clear_entrance_cache(self):
-        self._cached_entrances = None
+    def get_entrances(self, player: Optional[int] = None) -> List[Entrance]:
+        if player is not None:
+            return list(self.regions.entrance_cache[player].values())
+        return list(itertools.chain(*(self.regions.entrance_cache[player].values()
+                                      for player in self.regions.entrance_cache)))
 
     def register_indirect_condition(self, region: Region, entrance: Entrance):
         """Report that access to this Region can result in unlocking this Entrance,
         state.can_reach(Region) in the Entrance's traversal condition, as opposed to pure transition logic."""
         self.indirect_connections.setdefault(region, set()).add(entrance)
 
     def get_locations(self, player: Optional[int] = None) -> List[Location]:
-        if self._cached_locations is None:
-            self._cached_locations = [location for region in self.regions for location in region.locations]
         if player is not None:
-            return [location for location in self._cached_locations if location.player == player]
-        return self._cached_locations
-
-    def clear_location_cache(self):
-        self._cached_locations = None
+            return list(self.regions.location_cache[player].values())
+        return list(itertools.chain(*(self.regions.location_cache[player].values()
+                                      for player in self.regions.location_cache)))
```

Comment:
This looks slow. I'm wondering if we should add a cache for this, or return an Iterator instead of a list?

The change to evermore (and other worlds) fixes the performance regression, but I'm still not convinced this is the best solution yet.

Reply:
unfortunately returning an iterator failed tests. earlier versions actually had this cached. Do we lose enough performance here that we want to use the memory to cache it?

Reply:
I guess another method is to offer iterator versions of these functions and use them where it works.

Reply:
Should be addressed now.